pkg/ctl/install/install.go

```diff
@@ -0,0 +1,435 @@
+	// Add
+	addCtx, addCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer addCtxCancel()
+	if err := runGitCmd(addCtx, cloneDir, "add", "--", fi.opts.gitFluxPath); err != nil {
+		return err
+	}
+
+	// Confirm there is something to commit, otherwise move on
+	diffCtx, diffCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer diffCtxCancel()
+	if err := runGitCmd(diffCtx, cloneDir, "diff", "--cached", "--quiet", "--", fi.opts.gitFluxPath); err == nil {
+		logger.Info("Nothing to commit (the repository contained identical manifests), moving on")
+		return nil
+	} else if  _, ok := err.(*exec.ExitError); !ok{
+		return err
+	}
+
+	// Commit
+	commitCtx, commitCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer commitCtxCancel()
+	args := []string{"commit",
+		"-m", "Add Initial Flux configuration",
+		fmt.Sprintf("--author=%s <%s>", fi.opts.templateParams.GitUser, fi.opts.templateParams.GitEmail),
+	}
+	if err := runGitCmd(commitCtx, cloneDir, args...); err != nil {
+		return err
+	}
+
+	// Push
+	pushCtx, pushCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer pushCtxCancel()
+	return runGitCmd(pushCtx, cloneDir, "push")
+}
+
+func (fi *fluxInstaller) applyManifestsInDir(ctx context.Context, dir string) error {
+	// NOTE: invoking kubectl is not ideal, but it's the less hairy way to do it until
+	//       server-side apply is available ( https://github.com/kubernetes/enhancements/issues/555 )
+	applyCtx, applyCtxCancel := context.WithTimeout(ctx, fi.opts.timeout)
+	defer applyCtxCancel()
+	kubeconfigFile, err := ioutil.TempFile(os.TempDir(), "eksctl-flux-install-kubeconfig")
+	if err != nil {
+		return fmt.Errorf("cannot find temporary kubeconfig file: %s", err)
+	}
+	kubeconfigPath := kubeconfigFile.Name()
+	kubeconfigFile.Close()
+	// defer os.Remove(kubeconfigFile.Name())
+	if _, err := kubeconfig.Write(kubeconfigPath, *fi.k8sConfig, true); err != nil {
+		return fmt.Errorf("cannot write to temporary kubeconfig file (%s): %s", kubeconfigFile.Name(), err)
+	}
+
+	kubectlCmd := exec.CommandContext(applyCtx, "kubectl", "apply", "-f", dir)
+	kubectlCmd.Env = []string{"KUBECONFIG="+kubeconfigFile.Name()}
+	kubectlCmd.Stdout = os.Stdout
+	kubectlCmd.Stderr = os.Stderr
+	return kubectlCmd.Run()
+}
+
+func runGitCmd(ctx context.Context, dir string, args ...string) error {
+	gitCmd := exec.CommandContext(ctx, "git", args...)
+	gitCmd.Stdout = os.Stdout
+	gitCmd.Stderr = os.Stderr
+	gitCmd.Dir = dir
+	return gitCmd.Run()
+}
+
+func getFluxManifests(params install.TemplateParameters, cs *kubernetes.Clientset) (map[string][]byte, error) {
+	params.AdditionalFluxArgs = []string{"--sync-garbage-collection", "--manifest-generation"}
+	manifests, err := install.FillInTemplates(params)
+	if err != nil {
+		return nil, err
+	}
+	// Verify if the namespace to install Flux already exists and otherwise add a manifest for it
+	_, err = cs.CoreV1().Namespaces().Get(params.Namespace, metav1.GetOptions{})
```

Comment:
I think talking to k8s should not be the responsibility of this function (it makes it harder to test). I would move the responsibility of deciding to add the namespace to the caller of this function, which it actually checking if the file is in the map so I would rewrite that `if ns is in map` to `if ns exists in k8s`. So:

In `run()` I would say:

```
...
if namespace ! exists in the cluster {
   nsManifest := getNamespaceManifest()
   createResourceSync(nsManifest)
}
manifests:= getFluxManifests()
...
```

Reply:
Done. Please let me know if it matches what you had in mind.